pilot/pkg/model/mesh.go

```diff
@@ -0,0 +1,785 @@
+type Mesh struct {
+
+	// allEndpoints is a map of the UID to the server instance. The UID is expected
+	// to be unique across the mesh. Mesh uses this map during updates from
+	// Controllers to locate the previous representation  of the endpoint and check
+	// for changes to its metadata that could result in changes to the subsetEndpoints.
+	allEndpoints map[string]*Endpoint
+
+	// subsetDefinitions holds the metadata, basically the set of lable name values
+	// that ought to match corresponding attributes of the endpoint.
+	// The actual subset of allEndpoints that match these subset definitions is
+	// stored in subsetEndpoints. Mesh uses this map for identifying config changes
+	// to the subset's metadata. The key of this map is the Subset name.
+	subsetDefinitions map[string]*route.Subset
+
+	// subsetEndpoints allows for look-ups from the Subset name to a set of Endpoint UIDs.
+	// The UIDs in the set are guaranteed to exist in allEndpoints and the
+	// corresponding LbEndpointis guaranteed to satisfy the labels in the corresponding
+	// Subset.
+	subsetEndpoints subsetEndpoints
+
+	// reverseAttrMap provides reverse lookup from label name > label value >
+	// endpointUIDSet. All Endpoints in the set are guaranteed to match the
+	// attribute with the label name and value. Mesh uses the map for quickly
+	// updating subsetEndpoints in response to subset label changes.
+	reverseAttrMap attributeValues
+
+	// reverseEpSubsets provides a reverse lookup from endpoint key to the set of
+	// subsets this endpoint matched to. All subset keys associated with the endpoint
+	// are guaranteed to exist in subsetDefinitions. Mesh uses this to reverse
+	// lookup all subsets impacted by changes to Endpoint metadata.
+	reverseEpSubsets endpointSubsets
+
+	// Mutex guards guarantees consistency of updates to members shared across
+	// threads.
+	mu sync.RWMutex
+}
+
+// Endpoint is a platform independent representation of a Mesh Endpoints that
+// uses Envoy v2 API's LbEndpoint as its internal implementation. It also
+// provides utility methods intended for platform specific Controllers.
+type Endpoint xdsapi.LbEndpoint
+
+// endpointUIDSet is a unique set of Endpoint UIDs.
+type endpointUIDSet map[string]bool
+
+// valueEndpointSet associates a label value with an endpointUIDSet.
+type valueEndpointSet map[string]endpointUIDSet
+
+// attributeValues is a reverse lookup map from label name > label value > endpointUIDs
+// that all have the same attribute matching the label name and label value.
+type attributeValues map[string]valueEndpointSet
+
+// subsetNameSet is a unique set of subset names.
+type subsetNameSet map[string]bool
+
+// endpointSubsets maps an endpoint UID to a set of matching subsets.
+type endpointSubsets map[string]subsetNameSet
+
+// subsetEndpoints maps an subset name to a set of matching endpoint UIDs.
+type subsetEndpoints map[string]endpointUIDSet
+
+// SubsetEvent is a pilot configuration event for notifying the specified
```

Comment:
It's not configuration event only. It can be a service registry even.

Reply:
I think the term Event is being given too much weight. It's just an enum to know whether the subset should be deleted, added or modified. Conflating this to an event queues or similar is a huge stretch which we should probably avoid.